bika/lims/browser/worksheet/views/analyses.py

```diff
@@ -71,216 +71,372 @@ def __init__(self, context, request):
-                    'after': {},
-                    'choices': {},
-                    'class': {},
-                    'state_class': 'state-empty',
-                    'allow_edit': [],
-                    'colspan': {'Pos':len(self.columns) + len(self.interim_fields)},
-                    'rowspan': {'Pos':1},
-                    'Pos': pos,
-                    'Service': '',
-                    'Attachments': '',
-                    'state_title': 's',})
-                item['replace'] = {
-                    'Pos': "<table width='100%' cellpadding='0' cellspacing='0'>" + \
-                            "<tr><td class='pos'>%s</td>" % pos + \
-                            "<td align='right'>&nbsp;</td></tr></table>",
-                    'select_column': '',
-                    }
-                items.append(item)
-
-        items = sorted(items, key=itemgetter('Pos'))
-
-        slot_items = {} # pos:[item_nrs]
-        for x in range(len(items)):
-            p = items[x]['Pos']
-            if p in slot_items:
-                slot_items[p].append(x)
-            else:
-                slot_items[p] = [x, ]
-        actual_table_position = -1
-        # The first item in items[this position] gets a rowspan for it's
-        # "Position" column, which spans all other table rows in this position.
-        for pos, pos_items in slot_items.items():
-            actual_table_position += 1
-            x = pos_items[0]
-            if pos in empties:
+        layout = layout and layout or []
+        # Map the analysis uids with their positions.
+        occupied = []
+        next_positions = {}
+        for item in layout:
+            uid = item.get('analysis_uid', '')
+            slot = int(item['position'])
+            occupied.append(slot)
+            position = next_positions.get(slot, 1)
+            str_position = "{:010}:{:010}".format(slot, position)
+            next_positions[slot] = position + 1
+            uids_positions[uid] = str_position
+
+        # Fill empties
+        last_slot = max(occupied)
+        empties = [num for num in range(1, last_slot) if num not in occupied]
+        for empty_slot in empties:
+            str_position = "{:010}:{:010}".format(empty_slot, 1)
+            uid = "empty-{}".format(empty_slot)
+            uids_positions[uid] = str_position
+
+        return uids_positions
+
+    def get_item_position(self, analysis_uid):
+        """
+        Returns a list with the position for the analysis_uid passed in within
+        the current worksheet in accordance with the current layout, where
+        the first item from the list returned is the slot and the second is
+        the position of the analysis within the slot.
+        :param analysis_uid: uid of the analysis the position is requested
+        :return: the position (slot + position within slot) of the analysis
+        :rtype: list
+        """
+        str_position = self.uids_strpositions.get(analysis_uid, '')
```

Comment:
I did not add this check because this scenario must never happen... and if somehow happens, then the only reason is that the dev was playing without his/her homework being done. So yes, I prefer a very big exception to get raised instead of obfuscating it.